R/data.table.R

```diff
@@ -1336,9 +1336,25 @@ replace_dot_alias = function(e) {
 
     if (is.data.table(jval)) {
       setattr(jval, 'class', class(x)) # fix for #64
-      if (haskey(x) && all(key(x) %chin% names(jval)) && suppressWarnings(is.sorted(jval, by=key(x))))  # TO DO: perhaps this usage of is.sorted should be allowed internally then (tidy up and make efficient)
+      # can jval be sorted by the same key as x? improved for #4498
+      get_shared_keys = function(jsub, jvnames, key) {
+        if (is.null(key)) return(NULL)
+        if (!jsub %iscall% "list") return(NULL)
+        jnames = as.character(Filter(is.name, jsub)[-1L])
+        key_idx = chmatch(key, jnames)
+        missing_keys = which(is.na(key_idx))
+        if (length(missing_keys) && missing_keys[1L] == 1L) return(NULL)
+        if (!length(missing_keys)) return(jvnames[key_idx])
+        return(jvnames[head(key_idx, missing_keys[1L] - 1L)])
+      }
+      shared_keys = get_shared_keys(jsub, jvnames, key(x))
+      if (is.null(irows) && !is.null(shared_keys)) {
+        setattr(jval, 'sorted', shared_keys)
+        # potentially inefficient backup -- check if jval is sorted by key(x)
+      } else if (haskey(x) && all(key(x) %chin% names(jval)) && suppressWarnings(is.sorted(jval, by=key(x)))) { # TO DO: perhaps this usage of is.sorted should be allowed internally then (tidy up and make efficient)
```

Comment:
I would have dropped this branch but there are a few tests relying on it, e.g. when `lapply(.SD, as.numeric)` is used to integer column, it can still be sorted by key, or when `irows` is there but it's an ordered subset.

Do we check anywhere else whether `irows` is an ordered subset? maybe an opportunity to leverage `ALTREP` to quickly determine `1:n` subsets are ordered.

@ColeMiller1 please have a look, I think the logic is improved but maybe could be further.

Reply:
I see this above:

```
orderedirows = .Call(CisOrderedSubset, irows, nrow(x))  # TRUE when irows is NULL (i.e. no i clause).
```

but only in the `!missingby` region

Reply:
I like your new fx. We may want to check for ```is.name(.SD)``` so we can pull out those SD vars as well. 

I think the ```is.sorted()``` could be avoided if:

1. This is a join; or
2. ```!length(forder(irows))```

I can work on it some this evening.